libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -152,6 +152,55 @@ string YulUtilFunctions::storeLiteralInMemoryFunction(string const& _literal)
 	});
 }
 
+string YulUtilFunctions::copyLiteralToStorageFunction(string const& _literal)
+{
+	string functionName = "copy_literal_to_storage_" + util::toHex(util::keccak256(_literal).asBytes());
+
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(slot) {
+				<body>
+			}
+		)");
+		templ("functionName", functionName);
+
+		if (_literal.size() >= 32)
+		{
+			size_t words = (_literal.length() + 31) / 32;
+			vector<map<string, string>> wordParams(words);
+			for (size_t i = 0; i < words; ++i)
+			{
+				wordParams[i]["offset"] = to_string(i);
+				wordParams[i]["wordValue"] = formatAsStringOrNumber(_literal.substr(32 * i, 32));
+			}
+			templ("body", Whiskers(R"(
+					<resizeArray>(slot, <length>)
+					let dstPtr := <dataArea>(slot)
+					<#word>
+						sstore(add(dstPtr, <offset>), <wordValue>)
+					</word>
+				)")
+				("resizeArray", resizeDynamicByteArrayFunction(*TypeProvider::bytesStorage()))
+				("dataArea", arrayDataAreaFunction(*TypeProvider::bytesStorage()))
+				("word", wordParams)
+				("length", to_string(_literal.size()))
+				.render());
+		}
+		else
+			templ("body", Whiskers(R"(
+					<resizeArray>(slot, <length>)
+					sstore(slot, add(<wordValue>, <encodedLen>))
```

Comment:
There is an `sstore` inside of resizeArray function, and combined with `sstore` just below, we have multiple `sstore`s to the same slot in case of short literal. Before, we copied literal to memory and then we did copy to storage. In `copyByteArrayToStorage` function, we do not call directly `resizeArray` but we do truncation there  which doesn't produce two `sstore`s. This leads to having more gas used when we assign literal to storage array.

And in `copy_byte_array_to_storage` test, we have multiple similar assignments `data = "abc";`

Reply:
Oh, even more `sstore`s are waisted here.
I'll try explaining what else happens. So, we have for example part of the test mentioned:
```
data = "abc";
if (!(readDataSlot() == zero)) return 2;
data = "1234567890123456789012345678901234567890123456789012345678901234567890";
```
data is resized from short to long byte array. So, `resizeArray` copies "abc" to new data location... That data is being overwritten as the array is just being overwritten here. I would suggest adding flag to `resizeArray` to distinguish type of resize we want to have.

It looks like this type of resize is working only for push. 

Nothing of this is done in copying byte array from memory to storage...